### PR TITLE
In source build failure warning

### PR DESCRIPTION
**Pull request by hippo91**
This PR is an answer to #3081 
It just adds a FATAL_ERROR in message with explanation in case of in-source build.

### CODE REVIEW

**Comment by dalg24** [comment]
OK to test

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,4 +1,10 @@
 
+# Disable in-source builds to prevent source tree corruption.
+if( "${CMAKE_SOURCE_DIR}" STREQUAL "${CMAKE_BINARY_DIR}" )
+  message( FATAL_ERROR "FATAL: In-source builds are not allowed.
```
Comment (jjwilke) [inline CMakeLists.txt:4]: I would do this as a single-line string and let CMake do the wrapping. This prints weirdly on my terminal.
Maybe redundant to say "FATAL"?

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,4 +1,10 @@
...
+                       You should create a separate directory for build files." )
+endif()
```
Comment (jjwilke) [inline CMakeLists.txt:6]: I actually don't care about case anymore, but in the past we had wanted to follow Trilinos use of upper case. @dalg24

File: CMakeLists.txt (hunk 1 above)
Comment (hippo91) [reply CMakeLists.txt:4]: On my terminal, **FATAL** appears only once. Thus in my case, i don't see any redundancy:
```
CMake Error at CMakeLists.txt:4 (message):
  FATAL: In-source builds are not allowed.  You should create a separate
  directory for build files.


-- Configuring incomplete, errors occurred!
```

File: CMakeLists.txt (hunk 2 above)
Comment (hippo91) [reply CMakeLists.txt:6]: I can't find any reference to **Trilinos** cmake coding style. Could you explain a little further?

**Pull request merged**